[AMD] Define an extract slice operation

hmalgewatta opened the pull request:
This commit introduces an extract_slice operation for AMD backend
to enable extracting slice of a tensor in registers without data exchange.
It enables breaking down large tiles of tensors into smaller ones
for better instruction interleaving and scheduling.

This can be useful for hiding global memory latency when a global
load/store can be efficiently split into several loads/stores to be
overlapped with compute fo attention.

- [x] I am not making a trivial change, such as fixing a typo in a comment.

- [x] I have written a PR description following these
  [rules](https://cbea.ms/git-commit/#why-not-how).

- [x] I have run `pre-commit run --from-ref origin/main --to-ref HEAD`.

- Select one of the following.
  - [x] I have added tests.
    - `/test` for `lit` tests
    - `/unittest` for C++ tests
    - `/python/test` for end-to-end tests
  - [ ] This PR does not need a test because `FILL THIS IN`.

- Select one of the following.
  - [ ] I have not added any `lit` tests.
  - [x] The `lit` tests I have added follow these [best practices](https://mlir.llvm.org/getting_started/TestingGuide/#filecheck-best-practices),
    including the "tests should be minimal" section. (Usually running Python code
    and using the instructions it generates is not minimal.)

antiagainst commented on python/test/unit/language/test_core.py:
```diff
@@ -5619,3 +5619,86 @@ def kernel(
     np.testing.assert_allclose(y_ref, to_numpy(y_tri), rtol=0.01)
+
+
+# -----------------------
```
Not sure we want to add the test here. We can add a `third_party/amd/python/test` directory.

antiagainst commented on python/test/unit/language/test_core.py:
```diff
@@ -5619,3 +5619,86 @@ def kernel(
+# test view slice
+# -----------------------
+
+view_layout = [
```
These two are only used in the following test? Can you put them in the test?

hmalgewatta replied:
They are used to parameterize the test so I left them outside the test

antiagainst replied:
ah okay makes sense.

antiagainst commented on python/test/unit/language/test_core.py:
```diff
@@ -5619,3 +5619,86 @@ def kernel(
+    ir = f"""
+    #blocked = {blocked_layout}
+    #view_layout = {view_layout}
+    """ + """
```
You can just put the whole IR in one string? Why the need to separate out and concat?

antiagainst commented on python/test/unit/language/test_core.py:
```diff
@@ -5619,3 +5619,86 @@ def kernel(
+    view = torch.empty((M_tile_size, N_tile_size), device=device, dtype=torch.float16)
+
+    kernel[(1, 1, 1)](x.data_ptr(), view)
+    test_result = torch.eq(x[M_tile_offset:M_tile_size + M_tile_offset, N_tile_offset:N_tile_offset + N_tile_size],
```
`torch.equal` gives you the final result without needing to explicitly call `.all()`?

antiagainst commented on third_party/amd/lib/TritonAMDGPUDialectToLLVM/TritonAMDGPUToLLVMPatterns.cpp:
```diff
@@ -1,9 +1,11 @@
                                         RewritePatternSet &patterns,
                                         PatternBenefit benefit) {
-  // TODO: Insert TrtionAMDGPU dialect patterns.
+  populateViewSliceOpTritonAMDGPUToLLVMPatterns(typeConverter, patterns,
```
`populateViewSliceOpToLLVMPatterns`.

antiagainst commented on third_party/amd/include/TritonAMDGPUToLLVM/PatternTritonAMDGPUToLLVM.h:
```diff
@@ -0,0 +1,17 @@
+#ifndef THIRD_PARTY_AMD_INCLUDE_TRITONAMDGPU_TO_LLVM_PATTERNS_AMDGPU_OP_TO_LLVM_H
```
No need for the `THIRD_PARTY_AMD_INCLUDE_` part.

antiagainst commented on third_party/amd/include/TritonAMDGPUToLLVM/PatternTritonAMDGPUToLLVM.h:
```diff
@@ -0,0 +1,17 @@
+#define THIRD_PARTY_AMD_INCLUDE_TRITONAMDGPU_TO_LLVM_PATTERNS_AMDGPU_OP_TO_LLVM_H
+
+#include "mlir/Conversion/LLVMCommon/TypeConverter.h"
+#include "triton/Analysis/AxisInfo.h"
```
The `AxisInfo.h` header is not needed here?

antiagainst commented on third_party/amd/include/TritonAMDGPUToLLVM/PatternTritonAMDGPUToLLVM.h:
```diff
@@ -0,0 +1,17 @@
+#include "mlir/Conversion/LLVMCommon/TypeConverter.h"
+#include "triton/Analysis/AxisInfo.h"
+
+using namespace mlir;
```
Please don't `use namespace` in header files--header files are meant to be included in other places and it will trigger we `use namespace` everywhere.

antiagainst commented on third_party/amd/include/Dialect/TritonAMDGPU/IR/TritonAMDGPUOps.td:
```diff
@@ -31,5 +31,64 @@ include "mlir/Dialect/LLVMIR/LLVMOpBase.td"
 include "mlir/Interfaces/InferTypeOpInterface.td"
 include "TritonAMDGPUDialect.td"
 include "TritonAMDGPUAttrDefs.td"
+include "mlir/Interfaces/SideEffectInterfaces.td" // Pure
```
Please sort these header files alphabetically.

antiagainst commented on third_party/amd/include/Dialect/TritonAMDGPU/IR/TritonAMDGPUOps.td:
```diff
@@ -31,5 +31,64 @@ include "mlir/Dialect/LLVMIR/LLVMOpBase.td"
+    Op<TritonAMDGPU_Dialect, mnemonic, traits>;
+
+def TritonAMDGPU_ViewSliceOp : TritonAMDGPU_Op<"view_slice",
+                                [AttrSizedOperandSegments,
```
Can you concatenate these into oneline and sort them alphabetically?

antiagainst commented on third_party/amd/include/Dialect/TritonAMDGPU/IR/TritonAMDGPUOps.td:
```diff
@@ -31,5 +31,64 @@ include "mlir/Dialect/LLVMIR/LLVMOpBase.td"
+                                 ]> {
+  let summary = "view slice operation";
+  let description = [{
+    Represents view of the slice of the tensor in registers. However, unlike 'memdesc_subview' which provides a view in shared memory,
```
We need to write better docs for this op's semantics; it's too vague right now. You can see how stuff are defined for https://mlir.llvm.org/docs/Dialects/TensorOps/#tensorextract_slice-tensorextractsliceop and https://mlir.llvm.org/docs/Dialects/MemRef/#memrefsubview-memrefsubviewop for basic descriptions. Then we can focus on explaining the additional requirements on registers and layout etc. 

Also please align it to 80 chars.

antiagainst commented on third_party/amd/lib/Dialect/TritonAMDGPU/IR/Dialect.cpp:
```diff
@@ -46,3 +47,49 @@ void mlir::triton::amdgpu::TritonAMDGPUDialect::initialize() {
+      dyn_cast<RankedTensorType>(resultTy).getElementType();
+
+  if (srcElementType != resultElementType) {
+    return emitError("result type must match source type");
```
"source element type" to be precise.

antiagainst commented on third_party/amd/lib/Dialect/TritonAMDGPU/IR/Dialect.cpp:
```diff
@@ -46,3 +47,49 @@ void mlir::triton::amdgpu::TritonAMDGPUDialect::initialize() {
+LogicalResult ViewSliceOp::verify() {
+  auto srcTy = getSource().getType();
+  auto srcLayout = srcTy.getEncoding();
+  auto srcElementType = dyn_cast<RankedTensorType>(srcTy).getElementType();
```
You can use `mlir::getElementTypeOrSelf` here to handle scalar cases. Similarly for `resultElementType`. Though--I think we should error out for scalar cases?

hmalgewatta replied:
The source input is supposed to be an AnyRankedTensor. I was thinking that would take care of scalar cases. Is that not the case?

antiagainst commented on third_party/amd/include/Dialect/TritonAMDGPU/IR/TritonAMDGPUOps.td:
```diff
@@ -31,5 +31,64 @@ include "mlir/Dialect/LLVMIR/LLVMOpBase.td"
+
+  let arguments = (ins
+    AnyRankedTensor:$source,
+    Variadic<I32>:$offsets,
```
Hmm, do we actually wanting to support dynamic offset/size/stride cases? Isn't this op closer to https://github.com/llvm/llvm-project/blob/6292f117c39b9fc72da4e40328eeeda2aa94a5f2/mlir/include/mlir/Dialect/Vector/IR/VectorOps.td#L1183

antiagainst replied:
What I meant is not 2D or 3D; I meant is whether we need to support dynamic offsets/sizes/strides; that is, something like `view_slice %source[%off0, %off1][%size0, %size1]..`. `OffsetSizeAndStrideOpInterface` supports that but I think we only need static offsets/sizes/strides given we are slicing in registers (well tensors in IR)?

@oplavsic wdyt?

hmalgewatta replied:
If we want to make this a more generalized operation we could add support for dynamic offsets/sizes/strides. That would require me to modify the current conversion to llvm and insert llvm instruction to dynamically calculate these. @oplavsic and I talked about this today, and I think for supporting dot-slicing static offsets//sizes/strides are sufficient.

antiagainst replied:
Yeah I'm fine with this now. But could you add a validation to error out for non-static cases at least?

hmalgewatta replied:
I could also change the instruction to have them only as attributes like https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/Dialect/Vector/IR/VectorOps.td#L1183

```
%72 = amdgpu.view_slice %arg0 {
  offsets = array<i64: 0,0>,
  sizes = array<i64: 256,16>,
  strides = array<i64: 1,1>
} : tensor<256x128xi32, #blocked1> to tensor<256x16xi32, #blocked1>
```

Is this better or should I keep the old format and verify for non-static?

antiagainst replied:
You can keep the old format and verify to reject non-static case for now.

antiagainst commented on third_party/amd/lib/Dialect/TritonAMDGPU/IR/Dialect.cpp:
```diff
@@ -46,3 +47,49 @@ void mlir::triton::amdgpu::TritonAMDGPUDialect::initialize() {
+
+  if (srcLayout != resultLayout)
+    return emitError("result layout must match source layout");
+
```
The below implicitly assumes rank == 2? Can we error out as unimplemented for other cases at least?

antiagainst commented on third_party/amd/lib/Dialect/TritonAMDGPU/IR/Dialect.cpp:
```diff
@@ -46,3 +47,49 @@ void mlir::triton::amdgpu::TritonAMDGPUDialect::initialize() {
+  }
+
+  if (sizes[0] % shapePerCTA[0] != 0 || sizes[1] % shapePerCTA[1] != 0) {
+    return emitError("incorrect static size");
```
Please be verbose regarding error messages--something like "slicing size X is not a multiple of shapePerCTA Y". Also please add lit test for failure cases. 

Similarly for the offset and stride errors.

antiagainst commented on third_party/amd/lib/TritonAMDGPUDialectToLLVM/ViewSliceOpToLLVM.cpp:
```diff
@@ -0,0 +1,138 @@
+  LogicalResult processLayout(tta::ViewSliceOp op, OpAdaptor adaptor,
+                              ConversionPatternRewriter &rewriter) const {
+    Location loc = op->getLoc();
+    auto srcTy = dyn_cast<RankedTensorType>(op.getSource().getType());
```
`dyn_cast` can return null so generally you need to handle the case it's not a `RankedTensorType`. I think you want `cast` here (and use the verifier to guarantee we don't have scalar cases).

antiagainst commented on third_party/amd/lib/TritonAMDGPUDialectToLLVM/ViewSliceOpToLLVM.cpp:
```diff
@@ -0,0 +1,138 @@
+#include "triton/Conversion/TritonGPUToLLVM/PatternTritonGPUOpToLLVM.h"
+#include "triton/Conversion/TritonGPUToLLVM/Utility.h"
+
+using namespace mlir;
```
Can we don't just bring all symbols from all namespaces in? The purpose of different namespaces is to avoid symbol collisons. This basically defies that purpose. :) It's fine to use the first two; please drop the last two.

antiagainst commented on third_party/amd/lib/TritonAMDGPUDialectToLLVM/ViewSliceOpToLLVM.cpp:
```diff
@@ -0,0 +1,138 @@
+    if (isa<BlockedEncodingAttr>(op.getSource().getType().getEncoding()) ||
+        isa<AMDMfmaEncodingAttr>(op.getSource().getType().getEncoding())) {
+      return processLayout(op, adaptor, rewriter);
+    } else {
```
No need to have the `else` given the above `return`s.

antiagainst commented on third_party/amd/lib/TritonAMDGPUDialectToLLVM/ViewSliceOpToLLVM.cpp:
```diff
@@ -0,0 +1,138 @@
+        isa<AMDMfmaEncodingAttr>(op.getSource().getType().getEncoding())) {
+      return processLayout(op, adaptor, rewriter);
+    } else {
+      assert(false && "Unsupported layout in viewSlice.");
```
Why `assert` here? Naturally failing the pattern rewrite with `return failure()` is enough?

antiagainst commented on third_party/amd/lib/TritonAMDGPUDialectToLLVM/ViewSliceOpToLLVM.cpp:
```diff
@@ -0,0 +1,138 @@
+    auto totalSizePerThread = sizePerThread[0] * sizePerThread[1];
+    auto order = getOrder(srcLayout);
+    auto shapePerCTA = getShapePerCTATile(srcLayout, srcShape);
+    shapePerCTA[0] =
```
Why do we need this? Can you add comment for it?

antiagainst commented on third_party/amd/lib/TritonAMDGPUDialectToLLVM/ViewSliceOpToLLVM.cpp:
```diff
@@ -0,0 +1,138 @@
+    auto sizes = op.getStaticSizes();
+
+    // Calculate offsets and sizes in terms of CTA units.
+    std::vector<long int> CTAOffsets{offsets[0] / shapePerCTA[0],
```
What about using `int64_t` as a precise type?

antiagainst reviewed (changes requested):
Thanks! I have a bunch of inling comments. Major issues include refine the semantics of the op and adding more docs/tests for it.

hmalgewatta commented:
Thanks for meticulously going through the code changes! I have a new commit to address the comments. And I have also moved the python test case to the new proposed location. 
I left some parts in that test that are not necessary right now (taken from the original test_core.py) but might be useful if more tests were to be added. Let me know if I should remove them

antiagainst commented on test/TritonGPU/amd/amd-viewslice-op.mlir:
```diff
@@ -0,0 +1,24 @@
+
+module attributes {"triton_gpu.compute-capability" = 0 : i32, "triton_gpu.num-ctas" = 1 : i32, "triton_gpu.num-warps" = 8 : i32, "triton_gpu.threads-per-warp" = 64 : i32} {
+  tt.func @basic_insert_slice_async_1d(%arg0: tensor<256x128xi32, #blocked1> {tt.divisibility = 16 : i32}) {
+    // CHECK: llvm.func @basic_insert_slice_async_1d
```
This should be done by adding `--verify` to the `RUN` line and then `expect-error`. See examples in mlir https://github.com/search?q=repo%3Allvm%2Fllvm-project+expected-error+language%3AMLIR&type=code&l=MLIR

antiagainst commented on test/TritonGPU/amd/amd-viewslice-op.mlir:
```diff
@@ -0,0 +1,24 @@
+  }
+}
+
+module attributes {"triton_gpu.compute-capability" = 0 : i32, "triton_gpu.num-ctas" = 1 : i32, "triton_gpu.num-warps" = 8 : i32, "triton_gpu.threads-per-warp" = 64 : i32} {
```
typically op syntax/semantics check should be in its own file. Can you move the above to its own file and name as something like viewslice-to-llvm.mlir under test/Conversion?

antiagainst commented on test/TritonGPU/amd/amd-viewslice-op.mlir:
```diff
@@ -0,0 +1,24 @@
+module attributes {"triton_gpu.compute-capability" = 0 : i32, "triton_gpu.num-ctas" = 1 : i32, "triton_gpu.num-warps" = 8 : i32, "triton_gpu.threads-per-warp" = 64 : i32} {
+  tt.func @basic_insert_slice_async_1d(%arg0: tensor<256x128xi32, #blocked1> {tt.divisibility = 16 : i32}) {
+    // CHECK: llvm.func @basic_insert_slice_async_1d
+    // CHECK: error: sizes [256, 2] must be a multiple of shapePerCTA [256, 16]
```
This only checks one errors. Can you also add negative tests for other op verification errors?

antiagainst commented on third_party/amd/include/Dialect/TritonAMDGPU/IR/TritonAMDGPUOps.td:
```diff
@@ -21,15 +21,90 @@
+                                     OffsetSizeAndStrideOpInterface, Pure]> {
+  let summary = "view slice operation";
+  let description = [{
+    The "view_slice" operation enables "viewing" a slice of a tensor in
```
Nit: no need to quote "view"--view is sort of a common concept well known in languages/compilers already.

antiagainst commented on third_party/amd/include/Dialect/TritonAMDGPU/IR/TritonAMDGPUOps.td:
````diff
@@ -21,15 +21,90 @@
+    Example 1:
+
+    ```mlir
+    %1 = triton_gpu.convert_layout %0 : tensor<128x128x!tt.ptr<f16>, #blocked>
````
Can you also be specific about the `#blocked` and `#blocked2` and mention the limitations on them?

antiagainst commented on third_party/amd/include/Dialect/TritonAMDGPU/IR/TritonAMDGPUOps.td:
````diff
@@ -21,15 +21,90 @@
+    ```
+    }];
+
+  let arguments = (ins AnyRankedTensor:$source, Variadic<I32>:$offsets,
````
Nit: place each argument in one line

antiagainst commented on third_party/amd/include/TritonAMDGPUToLLVM/PatternTritonAMDGPUToLLVM.h:
```diff
@@ -0,0 +1,14 @@
+#ifndef AMD_INCLUDE_TRITONAMDGPU_TO_LLVM_PATTERNS_AMDGPU_OP_TO_LLVM_H
```
No need to have the `AMD_INCLUDE` prefix.

antiagainst commented on third_party/amd/python/test/test_core.py:
```diff
@@ -0,0 +1,146 @@
+# flake8: noqa: F821,F841
+import contextlib
```
Don't think we need all the `import`s here? Can you only retain necessary ones?

antiagainst commented on third_party/amd/python/test/test_core.py:
```diff
@@ -0,0 +1,146 @@
+num_ctas_list = [1]
+
+GPU_DIALECT = "triton_gpu"
+if is_interpreter():
```
We only care aboutt the hip case.

antiagainst reviewed (changes requested):
Thanks, much better!

antiagainst commented on third_party/amd/python/test/test_core.py:
```diff
@@ -0,0 +1,116 @@
+# flake8: noqa: F821,F841
```
Is this still needed?

antiagainst commented on third_party/amd/python/test/test_core.py:
I think we may want to rename this test into something like `test_view_slice.py` to be clear.

Also this test is not checked by the CI yet? Can you enable in CI?

antiagainst reviewed (changes requested):
Great! Some final comments inlined. Also could you `git merge origin/main` so that I can trigger CI? Right now cannot due to not using latest `main`.

oplavsic commented:
@hmalgewatta thanks for working on this and @antiagainst thanks for detailed review and feedback. Looks great now :)

hmalgewatta commented:
Hi @antiagainst in the most recent commit I renamed the test file, added code checking for non-static cases, added more lit tests for failing non-static cases and changes to avoid conflicts. I've also synced my fork with the main branch so that you can trigger CI

antiagainst commented on third_party/amd/lib/Dialect/TritonAMDGPU/IR/Dialect.cpp:
```diff
@@ -46,3 +47,73 @@ void mlir::triton::amdgpu::TritonAMDGPUDialect::initialize() {
+  shapePerCTA[1] = std::min(static_cast<unsigned>(srcShape[1]), shapePerCTA[1]);
+
+  auto checkForConstInts = [](Value val) {
+    return isa<arith::ConstantIntOp>(val.getDefiningOp());
```
There are two static cases: 1) static attributes baked into `DenseI64ArrayAttr:$static_offsets`, 2) `mlir::Value` that are constant referenced by `Variadic<Value>::I32`. This supports the second case but not the first? I believe you need something like `all_of(getMixedOffsets()), [](OpFoldResult ofr) { return getConstantIntValue(ofr).has_value(); })`.

antiagainst reviewed (changes requested):
Please `git merge origin/main` again and resolve the conflicts.

ThomasRaoux commented on third_party/amd/include/Dialect/TritonAMDGPU/IR/TritonAMDGPUOps.td:
```diff
@@ -21,21 +21,111 @@
+    }];
+
+  let arguments = (ins AnyRankedTensor:$source,
+      Variadic<I32>:$offsets,
+      Variadic<I32>:$sizes,
+      Variadic<I32>:$strides,
+      DenseI64ArrayAttr:$static_offsets,
+      DenseI64ArrayAttr:$static_sizes,
+      DenseI64ArrayAttr:$static_strides);
```
Do we really want to support dynamic indexes? Also I don't really understand the meaning of sizes and strides for a tensor slice as this will create a totally new immutable tensor, it sounds like this mixes up concept of a memory view.
I would actually suggest removing `view` from the op name as it is a bit confusing in the context of tensor. It feels like this is meant to implement a slice.

antiagainst replied:
Yup. This should be addressed now.

ThomasRaoux reviewed (commented):
The design of the op is significantly different than what I would have imagined so maybe I'm missing some context.
Let me know if my comment makes sense otherwise maybe we need to discuss this a bit more

antiagainst commented on third_party/amd/lib/Dialect/TritonAMDGPU/IR/Dialect.cpp:
```diff
@@ -47,3 +50,63 @@ void mlir::triton::amdgpu::TritonAMDGPUDialect::initialize() {
+                       << "] must be a multiple of shapePerCTA [" << shapePerCTA
+                       << "]";
+  }
+
```
Should we also add another check to make sure `offset[i] + size[i] <= srcTy.getDimSize(i)`?

hmalgewatta replied:
I added this check and also a lit test for this in my most recent commit

ThomasRaoux commented on third_party/amd/lib/TritonAMDGPUDialectToLLVM/ExtractSliceOpToLLVM.cpp:
```diff
@@ -0,0 +1,143 @@
+    for (auto i = 0; i < 2; ++i) {
+      sizes.push_back(resultTy.getDimSize(i));
+    }
+
+    auto offsets = op.getStaticOffsets();
+
+    // Calculate offsets and sizes in terms of CTA units.
+    std::vector<int64_t> CTAOffsets{offsets[0] / shapePerCTA[0],
+                                    offsets[1] / shapePerCTA[1]};
+    std::vector<int64_t> CTASizes{sizes[0] / shapePerCTA[0],
+                                  sizes[1] / shapePerCTA[1]};
+    std::vector<int64_t> CTAPerShape{srcShape[0] / shapePerCTA[0],
+                                     srcShape[1] / shapePerCTA[1]};
+
+    // The diagram above illustrates the graphical representation of the
+    // skipElems, tensorStride, and lastIdx variables.
+    auto skipElems = CTAOffsets[order[1]] *
+                         (elemsPerThread[order[0]] * sizePerThread[order[1]]) +
+                     CTAOffsets[order[0]] * totalSizePerThread;
+    auto tensorStride =
+        (CTAPerShape[order[0]] - CTASizes[order[0]]) * totalSizePerThread;
+    auto lastIdx =
+        (CTAOffsets[order[1]] + CTASizes[order[1]] - 1) *
+            elemsPerThread[order[0]] * sizePerThread[order[1]] +
+        (CTAOffsets[order[0]] + CTASizes[order[0]]) * totalSizePerThread;
+
+    assert(lastIdx <= vals.size());
+
+    SmallVector<Value> resultVals;
+    for (int i = skipElems; i < lastIdx; i += tensorStride) {
+      for (int j = 0; j < totalSizePerThread * CTASizes[order[0]]; ++j, ++i) {
+        assert(i < lastIdx);
+        resultVals.push_back(vals[i]);
+      }
+    }
+    Value ret = packLLElements(loc, this->getTypeConverter(), resultVals,
+                               rewriter, resultTy);
+
+    rewriter.replaceOp(op, ret);
+    return success();
```
this lowering makes assumptions on the layout and shape of the of the operands/destination that are strong than what is in the verifier right?
Where do we check that those are true. It is okay to fail lowering if we don't want to support some cases but we never want to miscompile.

antiagainst replied:
Good catch! +1. We should check that each thread is still holding the same elements in the op verifier.

hmalgewatta replied:
@antiagainst the additional constraints I've identified to be added to verifier is making sure the divisors are not zero. I think moving the assert (line 107) would create duplicated code in the verifier, but I can move it there if that's the preference. I'm stuck on how to check that elements each thread is holding is as same as the verifier. Could you point me to how this could be done? And also if there's any other checks I might have missed

antiagainst replied:
Sorry for the late reply; gotten distracted previously.. After addressing https://github.com/triton-lang/triton/pull/4804#discussion_r1815799290, looks the current logic should be enough to guarantee that after slicing, threads are handling slides of the original elements without exchange/duplication. So this reads fine to me now @ThomasRaoux. Let me know if you still think some parts are missing.

antiagainst commented on third_party/amd/lib/TritonAMDGPUDialectToLLVM/ExtractSliceOpToLLVM.cpp:
```diff
@@ -0,0 +1,143 @@
+  matchAndRewrite(amdgpu::ExtractSliceOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    auto srcTy = op.getSource().getType();
+    if (isa<BlockedEncodingAttr>(op.getSource().getType().getEncoding()) ||
```
`isa` can support multiple attributes in `<...>`.

ThomasRaoux reviewed (approved):
> So this reads fine to me now @ThomasRaoux. Let me know if you still think some parts are missing.

Yes I think this is fine. I got thrown off by some of the naming in the verifier but I think it is correct

ThomasRaoux commented on third_party/amd/lib/Dialect/TritonAMDGPU/IR/Dialect.cpp:
```diff
@@ -47,3 +50,72 @@ void mlir::triton::amdgpu::TritonAMDGPUDialect::initialize() {
+  }
+
+  auto srcShape = srcTy.getShape();
+  auto shapePerCTA = mlir::triton::gpu::getShapePerCTATile(srcLayout, srcShape);
```
I thought I commented on this but I don't see it anymore sorry if it is a duplicate. Can we find a better name for this variable. In the rest of the code shapePerCTA has a very different meaning, here you are only getting one "tile" of the shape per CTA, shapePerCTA means the sub tensor owned by a CTA

antiagainst replied:
Good catch! This should be fixed now after renaming to `shapePerCTATile`.

hmalgewatta commented:
Sorry I pushed a version that was not well merged. I'll correct
